proposals/exception-handling/Exceptions-formal-overview.md

```diff
@@ -0,0 +1,203 @@
+Similarly, to verify that the `rethrow l` instruction refers to a label surrounding the instructions of a catch block (call this a catch-label), we allow the `kind` attribute of labels in the validation context to be set to `catch` when the label is a catch-label.
+```
+labelkind ::= try | catch
+labeltype ::= {result resulttype, kind labelkind?}
+C ::= {..., labels labeltype}
+```
+The original notation `labels [t*]` is now an abbreviation for:
+
+```
+labels {result [t*], kind ε}
+```
+
+
+### Instructions
+
+
+```
+C.tags[x] = [t*]→[]
+-----------------------------
+C ⊢ throw x : [t1* t*]→[t2*]
+
+
+C.labels[l].kind = catch
+----------------------------
+C ⊢ rethrow l : [t1*]→[t2*]
+
+
+C.types[bt] = [t1*]→[t2*]
+C, labels {result [t2*], kind try} ⊢ instr* : [t1*]→[t2*]
+(C.tags[x] = [t*]→[] ∧
+ C, labels { result [t2*], kind catch } ⊢ instr'* : [t*]→[t2*])*
+(C, labels { result [t2*], kind catch } ⊢ instr''* : []→[t2*])?
+-----------------------------------------------------------------------------
+C ⊢ try bt instr* (catch x instr'*)* (catch_all instr''*)? end : [t1*]→[t2*]
+
+
+C.types[bt] = [t1*]→[t2*]
+C, labels {result [t2*], kind try} ⊢ instr* : [t1*]→[t2*]
+C.labels[l].kind = try
+----------------------------------------------------------
+C ⊢ try bt instr* delegate l : [t1*]→[t2*]
+```
+
+## Execution (Reduction)
+
+### Runtime structure
+
+Stores
+
+```
+S ::= {..., tags taginst*}
+```
+
+Tag Instances
+
+```
+taginst ::= {type tagtype}
+```
+
+Module Instances
+
+```
+m ::= {..., tags a*}
+```
+
+Administrative Instructions
+
+```
+instr ::= ... | throw a | catch_n{a? instr*}* instr* end
+        | delegate{l} instr* end | caught_m{a val^n} instr* end
+```
+
+Block contexts and label kinds
+
+So far block contexts are only used in the reduction of `br l` and `return`, and only include labels or values on the stack above the hole. If we want to be able to break jumping over try-catch and try-delegate blocks, we must allow for the new administrative control instructions to appear after labels in block contexts, mirroring the label kinds of labels in validation contexts.
+
+```
+B^k ::= val* B'^k instr*
+B'^0 ::= [_]
+B'^{k+1} ::= label_n{instr*} B^k end | catch_m{a? instr*}* B^{k+1} end | caught_m{a val*} B^{k+1} end | delegate{l} B^{k+1} end
+```
+
+Note that `label_n{instr*} label_kind? [_] end? end` could be seen as a simplified control frame.
+
+(Alternatively, we could have the above `label_kind`s be also labels,  remove the additional `label_m` from the execution rules below, and remove the execution rules below where the new administrative instructions only contain `val*`. This would make labels even more similar to control frames.)
+
+Throw Contexts
+
+```
+T ::= val* [_] instr* | label_n{instr*} T end | caught_m{a val^n} T end
+   | frame_n{F} T end
+```
+
+### Instructions
+
+
+```
+F; throw x  ↪  F; throw a  (if F.module.tagaddrs[x]=a)
+
+caught_m{a val^n} B^l[rethrow l] end
+  ↪ caught_m{a val^n} B^l[val^n (throw a)] end
```

Comment:
I think this doesn't affect the JS API, WDYT @Ms2ger ?

Reply:
Do you mean that both `B^l` should be `T`? The intention is that the `l`in `B^l` is matching the one in`rethrow l`, otherwise how will we know which `caught` to target?`Did I miss a spec change on this?

Reply:
Can we consider this resolved, or am I missing the error?